### PR TITLE
Update company property specs

### PR DESCRIPTION
**Pull request by SViccari**
Ticket: [CompanyProperties: Test expectation isn't being executed](https://trello.com/c/zp5S0QH9)

---------------
**Suggestion for those reviewing:** Filter out the `.yml` files.
screen shot 2018-12-12 at 4 47 41 pm

---------------
**Summary:**
This ticket experienced severe scope creep as I dove into the existing tests for CompanyProperties. The original goal was to address to particular test that wasn't executing the expectation because the response was returning an empty list. While updating this test to returns a non-empty response from the API, the API response was returning over 1000 results, resulting in a VCR file that was over 1MB. The response is so large because we're using the `demo` account which is used by everyone testing out the HubSpot API.

To resolve this concern, this PR pulls in the `dotenv` library so the `hapikey` can be set to a value other than `demo`. In my case, I followed HubSpot's documentation for creating a [Developer Account] and used the Portal hapikey. This resulted in reducing the VCR recording that was 1.3MB down to 136KB. With time, I think we'll want to replace using the `hapikey` with an OAuth Access Token, as recommended by the HubSpot in the  Authentication API docs.

[Developer Account]: https://developers.hubspot.com/docs/methods/auth/oauth-overview

**A few other important items this PR addresses:**
- Removes trailing slashes from the CompanyProperty paths (I noticed this because VCR was recording requests to `/companies/v2/properties/?hapikey=<HAPI_KEY>` instead of `/companies/v2/properties?hapikey=<HAPI_KEY>`
- Updates VCR to filter out the HAPIKEY so it's not stored in the recording
- Updates `Properties#valid_property_params` and `Properties#valid_property_params` to treat string and symbol keys as the same.

### CODE REVIEW

File: spec/lib/hubspot/company_properties_spec.rb [hunk 1]
```diff
@@ -14,158 +14,325 @@
...
+
+            response = Hubspot::CompanyProperties.all({}, { include: [group_a] })
+
+            assert_hubspot_api_request(:get, "/companies/v2/properties")
+
+            group_names = response.map { |property| property["groupName"] }
+            expect(group_names).to include(group_a)
+            expect(group_names).not_to include(group_b)
+
+            Hubspot::CompanyProperties.delete!("instagram_handle")
```
Comment (cbisnett) [inline spec/lib/hubspot/company_properties_spec.rb]: If the `assert` or any of the `expect`'s above fail do these get cleaned up? If not, does that break the test next iteration when the properties are created?

These may be better in a `before` and `after` block.

File: spec/lib/hubspot/company_properties_spec.rb [hunk 2]
```diff
@@ -14,158 +14,325 @@
     end
   end
 
-  let(:example_groups) do
-    VCR.use_cassette('company_properties/groups_example') do
-      HTTParty.get('https://api.hubapi.com/companies/v2/groups?hapikey=demo').parsed_response
-    end
-  end
+  before { Hubspot.configure(hapikey: ENV.fetch("HUBSPOT_HAPI_KEY", "demo")) }
```
Comment (cbisnett) [inline spec/lib/hubspot/company_properties_spec.rb]: This is something I've been meaning to do for a while in addition to having a test application setup for tests run in Travis CI. This way the tests won't break if someone is doing something funky in the demo application. We should set this up for all tests possibly in the RSpec configuration block.

File: spec/lib/hubspot/company_properties_spec.rb [hunk 3]
```diff
@@ -14,158 +14,325 @@
...
+              { exclude: group_a }
+            )
+
+            assert_hubspot_api_request(:get, "/companies/v2/groups")
+
+            group_names = response.map { |group| group["name"] }
+            expect(group_names).not_to include(group_a)
+            expect(group_names).to include(group_b)
+
+            Hubspot::CompanyProperties.delete!("instagram_handle")
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Will these get called if the test fails? Should these be in an after hook or something similar to ensure they happen?

File: spec/lib/hubspot/company_properties_spec.rb [hunk 4]
```diff
@@ -14,158 +14,325 @@
...
+            })
+
+            response = Hubspot::CompanyProperties.groups(
+              {},
+              { include: group_a }
+            )
+
+            assert_hubspot_api_request(:get, "/companies/v2/groups")
+
+            Hubspot::CompanyProperties.delete!("instagram_handle")
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Same question here around ensuring that these happen even if the exercise or expectation part of this test fails.

File: spec/lib/hubspot/company_properties_spec.rb [hunk 5]
```diff
@@ -14,158 +14,325 @@
...
-        it 'should return nil' do
-          expect(Hubspot::CompanyProperties.delete!(name)).to eq(nil)
+          response = Hubspot::CompanyProperties.delete!(name)
+
+          assert_hubspot_api_request(
+            :delete,
+            "/companies/v2/properties/named/#{name}"
+          )
+
+          expect(response).to be nil
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Is it worth having an expectation on the status code?

File: spec/lib/hubspot/company_properties_spec.rb [hunk 6]
```diff
@@ -14,158 +14,325 @@
...
-          params['description']       = 'What is their favorite flavor?'
-          valid_params['description'] = params['description']
+          response = Hubspot::CompanyProperties.update!(
+            company_property_name,
+            params
+          )
 
-          response = Hubspot::CompanyProperties.update!(params['name'], params)
-          valid_params.each { |k, v| expect(response[k]).to eq(v) }
+          expect(response).to be nil
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Same here, is it worth expecting the status code to be something?

File: spec/lib/hubspot/company_properties_spec.rb [hunk 7]
```diff
@@ -14,158 +14,325 @@
...
+            body: {
+              groupName: "companyinformation",
+              label: new_label,
+              options: [],
+              type: "string",
+            }
+          )
+          expect(response["label"]).to eq(new_label)
+
+          Hubspot::CompanyProperties.delete!(name)
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: This is another clean up step that might not happen if the test fails.

File: spec/lib/hubspot/company_properties_spec.rb [hunk 8]
```diff
@@ -14,158 +14,325 @@
...
-      context 'with all valid parameters' do
-        cassette 'company_properties/create_property'
+      context "with no valid parameters" do
+        it "returns nil" do
+          response = Hubspot::CompanyProperties.create!({})
 
-        it 'should return the valid parameters' do
-          response = Hubspot::CompanyProperties.create!(params)
-          valid_params.each { |k, v| expect(response[k]).to eq(v) }
+          expect(response).to be nil
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Can we expect anything else here?

File: spec/lib/hubspot/company_properties_spec.rb [hunk 9]
```diff
@@ -14,158 +14,325 @@
...
+          )
+          expect(response["name"]).to eq(name)
+
+          Hubspot::CompanyProperties.delete!(name)
         end
       end
 
-      context 'with all valid parameters' do
-        cassette 'company_properties/create_property'
+      context "with no valid parameters" do
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Now that our expectation doesn't check the params, should we rename this test?

File: spec/lib/hubspot/company_properties_spec.rb [hunk 10]
```diff
@@ -14,158 +14,325 @@
...
+            "/companies/v2/properties",
+            body: {
+              name: name,
+              groupName: "companyinformation",
+              options: []
+            }
+          )
+          expect(response["name"]).to eq(name)
+
+          Hubspot::CompanyProperties.delete!(name)
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Is it possible this clean up step gets missed?

File: spec/lib/hubspot/company_properties_spec.rb [hunk 11]
```diff
@@ -14,158 +14,325 @@
...
+
+            response = Hubspot::CompanyProperties.all({}, { exclude: [group_a] })
+
+            assert_hubspot_api_request(:get, "/companies/v2/properties")
+
+            group_names = response.map { |property| property["groupName"] }
+            expect(group_names).not_to include(group_a)
+            expect(group_names).to include(group_b)
+
+            Hubspot::CompanyProperties.delete!("fax_number")
```
Comment (MattMSumner) [inline spec/lib/hubspot/company_properties_spec.rb]: Same question here about cleanup.

File: lib/hubspot/properties.rb [hunk 12]
```diff
@@ -88,7 +88,7 @@ def filter_results(results, key, include, exclude)
 
       def valid_property_params(params)
         return {} if params.blank?
-        result = params.slice(*PROPERTY_SPECS[:field_names])
+        result = params.with_indifferent_access.slice(*PROPERTY_SPECS[:field_names])
```
Comment (MattMSumner) [inline lib/hubspot/properties.rb]: Is this change required only for the tests to pass? Would it be possible to leave this as is and in our tests pass in a hash `with_indifferent_access`?

Comment (SViccari) [reply lib/hubspot/properties.rb]: > Is this change required only for the tests to pass?

It's not required for the tests to pass. I bumped into this when writing tests that pass in a hash with symbol keys instead of string keys. It wasn't obvious to me as to why the request was failing and it took time to realize the keys were being removed because they didn't match the `PROPERTY_SPECS[:field_names]` string keys.

> Would it be possible to leave this as is and in our tests pass in a hash `with_indifferent_access`

Certainly. However, I worry users will encounter the same issue I did (should the user pass in a hash with symbol keys) and spend time digging through this project's source code to figure out why their request is failing. For that reason, I think this change is worth keeping.

Ideally, I think this library doesn't need to be an arbiter of valid field names. That responsibility should fall on the API and any error messages returned by the API will be returned to the library user. 

@cbisnett Do you have any concerns about removing the logic that validates field names?

File: spec/lib/hubspot/company_properties_spec.rb (hunk 1 above)
Comment (SViccari) [reply spec/lib/hubspot/company_properties_spec.rb]: > If the assert or any of the expect's above fail do these get cleaned up?

They do not. The test will exit before running these cleanup actions. That shouldn't be an issue most of the time except when the VCR recording is being updated. I'm still finding the best way to work with VCR and currently, my flow is to write the API calls, capture all the successful calls in a recording, and then write expectations against the recording.

 > If not, does that break the test next iteration when the properties are created?

Yes 😞 I have run into this where I create a record, the update or next step fails because the API behaves differently than I expected, and I then need to delete the created record before I can try to record the API interactions again. I don't know of a great way around this as we want each test to be self-sufficient and that means sometimes having to remove records manually. I use Postman when I need to remove a record.

> These may be better in a before and after block.

I'll play around with this and report back.

File: spec/lib/hubspot/company_properties_spec.rb (hunk 2 above)
Comment (SViccari) [reply spec/lib/hubspot/company_properties_spec.rb]: I think we can avoid setting up a test application for Travis CI. The tests, when ran by Travis or ran locally with VCR_RECORD turned off, we can use a fake key value, like `fake-123`.

I agree it would be great to avoid using the "demo" account when recording API interactions as the demo account returns more data than we want to store in a VCR file. 

I'd like to update the tests to use the Oauth flow as the HubSpot API docs recommend using an access token from the OAuth flow as the preferred way to authenticate. I don't know if I'll have time to reach this goal but just mentioning it here as food for thought.

File: spec/support/hubspot_api_helpers.rb [hunk 13]
```diff
@@ -1,6 +1,10 @@
 module HubspotApiHelpers
   def hubspot_api_url(path)
-    URI.join(Hubspot::Config.base_url, path)
+    URI.join(Hubspot::Config.base_url, path).to_s
```
[the remark this answers is not in the document]
Comment (SViccari) [reply spec/support/hubspot_api_helpers.rb]: It can be removed 👍

File: spec/lib/hubspot/company_properties_spec.rb (hunk 3 above)
Comment (SViccari) [reply spec/lib/hubspot/company_properties_spec.rb]: I left an answer to a similar question here: https://github.com/adimichele/hubspot-ruby/pull/159#discussion_r241551906

Thinking about this a bit more, I could do something that calls these lines and rescues in the event the call fails. 

The benefit is the test author doesn't have to use a tool, like Postman, to delete the created record before they can re-run the test.

The downside to this change is the failure to delete a record is hidden. That's probably fine but it means a test author could commit a VCR recording where the `delete` action is a non-200 response. This could lead to some confusion when attempting to update this test's recording and the test errors due to a validation error. As we move away from using the "demo" account, that will become less of a likely scenario.

File: spec/lib/hubspot/company_properties_spec.rb (hunk 5 above)
Comment (SViccari) [reply spec/lib/hubspot/company_properties_spec.rb]: This method is returning just `nil`, so I would need to alter the return value of this function to check the status code. I think that's worth doing but I'd prefer to make that change in a separate PR.

File: spec/lib/hubspot/company_properties_spec.rb (hunk 6 above)
Comment (SViccari) [reply spec/lib/hubspot/company_properties_spec.rb]: The `update!` function returns early with `nil` when given empty params. An API call is not issued.

File: spec/lib/hubspot/company_properties_spec.rb (hunk 9 above)
Comment (SViccari) [reply spec/lib/hubspot/company_properties_spec.rb]: The params are still being checked. `create!` will return early when the params are empty. But to your point, I can update this to be more specific. I'll update this to read `with empty params`.

File: spec/lib/hubspot/company_properties_spec.rb (hunk 1 above)
Comment (SViccari) [reply spec/lib/hubspot/company_properties_spec.rb]: I attempted to implement an `around` hook to handle the test setup and cleanup. Something like:
```
RSpec.describe CompanyProperties do
  describe ".all" do
    context "context" do
      around(:example) do |example|
        ... create data in HubSpot ...
        example.run
        .... delete data in HubSpot ...
      end

      it "runs" do
        ... assertion ...
      end
    end 
  end
end
```

This worked well for 99% of the tests in the CompanyProperties spec file but failed intermittently for one test. When run after the `.all` tests, the `delete!` test failed with the error message that VCR expected one HTTP request to the delete endpoint but found two calls being made. I'm fairly certain this had to do with the use of the `around` blocks but I couldn't identify the exact reason. I'm not sure if it's an RSpec issue or a VCR issue.

Since this approach hasn't worked (so far), an alternative is to do something like:
```
  ... destroy record that's about to be created and rescue if the record doesn't exist ...
  ... create record ...
```

This changes the test cadence from `test setup, run test, remove records created in test setup` to `delete any data that will conflict with test setup, test setup, run test`.

For now, I'm going to merge and keep moving as we continue to find the best way to test with VCR.
